CefGlue.Common.Shared/Serialization/JsonConverters/ObjectJsonConverter.cs

```diff
@@ -1,60 +1,207 @@
+
+                    case JsonTokenType.EndObject:
+                        stateStack.Pop();
+                        continue;
+                }
+
+                SetObjectPropertyOrArrayIndex(currentState, value);
+            } while (reader.Read());
+
+            if (stateStack.Count > 1 && stateStack.Pop().ObjectHolder != root)
+            {
+                throw new InvalidOperationException();
+            }
+
+            return root.First();
+        }
 
-                default:
-                    return JsonDocument.ParseValue(ref reader).RootElement.Clone();
+        private static void SetObjectPropertyOrArrayIndex(
+            ReadState state,
+            object value)
+        {
+            var obj = state.ObjectHolder;
+            if (state.PropertyName != null)
+            {
+                ((Dictionary<string, object>)obj)[state.PropertyName] = value;
+            }
+            else if (state.ArrayIndex != null)
+            {
+                ((object[])obj)[state.ArrayIndex.Value] = value;
+                state.ArrayIndex++;
+            }
+        }
+
+        private bool HandleReferences(ref Utf8JsonReader reader, Dictionary<string, object> referencesMap, out object value, out bool isNewArray)
+        {
+            isNewArray = false;
+            var tempReader = reader;
+            tempReader.ReadToken(JsonTokenType.StartObject);
+            var propName = tempReader.ReadPropertyName();
+
+            var isId = propName == JsonAttributeIdPropName;
+            var isRef = !isId && propName == JsonAttributeRefPropName;
+            if (!isId && !isRef)
+            {
+                value = null;
+                return false;
+            }
+
+            reader.Read(); // StartObject
+            reader.Read(); // PropertyName
+            var objId = reader.GetString();
+
+            if (isRef)
+            {
+                value = referencesMap[objId];
+                return true;
+            }
+
+            // it's $id="N"
+            // check if it's followed by $values=[...]
+            tempReader.Read(); // $id Value
+            propName = tempReader.ReadPropertyName();
+            if (propName == JsonAttributeValuesPropName)
+            {
+                reader.Read(); // objId
+                reader.Read(); // $values PropertyName
+                value = new object[PeekAndCalculateArraySize(reader)];
+                isNewArray = true;
+            }
+            else
+            {
+                value = new Dictionary<string, object>();
+            }
+            referencesMap.Add(objId, value);
+            return true;
+        }
+
+        private int PeekAndCalculateArraySize(Utf8JsonReader reader)
+        {
+            var arraySize = 0;
+            var startDepth = reader.CurrentDepth;
+            reader.ReadToken(JsonTokenType.StartArray);
+            while (reader.CurrentDepth > startDepth || reader.TokenType != JsonTokenType.EndArray)
```

Comment:
I think it reads better because the stop condition is when the current depth == start and we found the endarray token
```suggestion
            while (!(reader.CurrentDepth == startDepth && reader.TokenType == JsonTokenType.EndArray))
```

Reply:
I actually find it harder to understand as it's a negation of a multiple steps condition. The current resolves in a simpler way in my mid:
Continue if condition 1 is true or condition 2 is true